File: benchmark/punctuation-restoration/parity/mojicast.py
"""Reference outputs for the Mojicast port: Mojicast's own punct.py under Python onnxruntime.

Imports punct.py straight from a clone of github.com/ishiki-emo/mojicast (@d06ddb9f) —
no re-implementation — and runs add_punctuation over the ja rows of results/inputs.json
at both precisions. The per-character sigmoid probabilities come from the same session
(punct._sess) with the same ids, so threshold flips between runtimes can be located.

usage: /home/jiangzhuo/.cache/sokuji-punct-bench/.venv/bin/python parity/mojicast.py
writes results/parity-mojicast-python.json
"""
import json
import os
import sys
import time

MOJICAST_SRC = "/home/jiangzhuo/.cache/sokuji-punct-bench/src/mojicast"
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, MOJICAST_SRC)

import numpy as np  # noqa: E402
import onnxruntime as ort  # noqa: E402
import punct  # noqa: E402


def probs_for(text):
    """Per-character [、, 。] sigmoid for the already-stripped text, chunked like _punctuate_raw."""
    res = []
    for i in range(0, len(text), 256):
        chunk = text[i:i + 256]
        ids = [punct._cls_id] + [punct._vocab.get(ch, punct._unk_id) for ch in chunk] + [punct._sep_id]
        a = np.array([ids], dtype=np.int64)
        logits = punct._sess.run(None, {"input_ids": a, "attention_mask": np.ones_like(a)})[0][0]
        p = 1.0 / (1.0 + np.exp(-logits))
        res.extend([[float(x) for x in p[j + 1]] for j in range(len(chunk))])
    return res


def main():
    rows = [r for r in json.load(open(os.path.join(ROOT, "results", "inputs.json"), encoding="utf-8"))
            if r["lang"] == "ja"]
    out = {"versions": {"onnxruntime": ort.__version__, "numpy": np.__version__}, "precisions": {}}
    for prec in ("int8", "fp32"):
        punct.unload()
        t0 = time.time()
        punct.load_punctuator(num_threads=4, precision=prec)   # includes the self-test
        load_s = time.time() - t0
        res = []
        for r in rows:
            text = r["input"].replace("、", "").replace("。", "")
            res.append({
                "id": r["id"], "variant": r["variant"], "input": r["input"],
                "out": punct.add_punctuation(r["input"]),
                "raw": punct._punctuate_raw(text) if text else text,
                "probs": probs_for(text),
            })
        out["precisions"][prec] = {"load_s": load_s, "probe": punct._punctuate_raw(punct._PROBE), "rows": res}
        print(prec, "load", round(load_s, 2), "probe", out["precisions"][prec]["probe"])
    a = out["precisions"]["int8"]["rows"]
    b = out["precisions"]["fp32"]["rows"]
    agree = sum(x["out"] == y["out"] for x, y in zip(a, b))
    out["int8_vs_fp32_exact"] = agree / len(a)
    print("python int8 vs fp32 exact:", agree, "/", len(a))
    for x, y in zip(a, b):
        if x["out"] != y["out"]:
            print("  int8:", x["out"])
            print("  fp32:", y["out"])
    path = os.path.join(ROOT, "results", "parity-mojicast-python.json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(out, f, ensure_ascii=False, indent=1)
    print("wrote", path)


if __name__ == "__main__":
    main()

File: benchmark/punctuation-restoration/parity/mojicast.mjs
// Compare models/mojicast.mjs and models/mojicast-fp32.mjs (onnxruntime-web WASM, 1 thread)
// with Mojicast's own punct.py under Python onnxruntime (parity/mojicast.py), and measure
// int8 vs fp32 agreement on both runtimes. Latency: parity/mojicast-latency.mjs.
// usage: node parity/mojicast.mjs   (after parity/mojicast.py)
// writes results/parity-mojicast.json
import { readFile, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { loadModel } from '../lib/node-env.mjs';
import { analyze } from '../lib/text.mjs';

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
const ref = JSON.parse(await readFile(join(root, 'results', 'parity-mojicast-python.json'), 'utf8'));
const THRESH = 0.1;
const rate = (n, d) => (d ? n / d : null);

/** Mark decided after each character: '' | '、' | '。' (before _NUM_PUNC_FIX). */
const decisions = (probs) => probs.map(([c, p]) => (p > THRESH ? '。' : c > THRESH ? '、' : ''));

function markAgreement(a, b) {
  let both = 0;
  let onlyA = 0;
  let onlyB = 0;
  let differ = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] && b[i]) (a[i] === b[i] ? both++ : differ++);
    else if (a[i]) onlyA++;
    else if (b[i]) onlyB++;
  }
  return { both, differentMark: differ, onlyFirst: onlyA, onlySecond: onlyB };
}

const report = { precisions: {}, pythonVersions: ref.versions };
const js = {};
for (const [id, prec] of [['mojicast', 'int8'], ['mojicast-fp32', 'fp32']]) {
  const { model, loadMs } = await loadModel(id, { threads: 1 });
  const py = ref.precisions[prec];
  const rows = [];
  let exact = 0;
  let rawExact = 0;
  let skeleton = 0;
  let maxAbs = 0;
  const flips = [];
  const diffs = [];
  for (const r of py.rows) {
    const j = await model.inspect(r.input);
    rows.push(j);
    exact += j.out === r.out;
    rawExact += j.raw === r.raw;
    skeleton += analyze(j.out).skel === analyze(r.input).skel;
    for (let i = 0; i < r.probs.length; i++) {
      for (let k = 0; k < 2; k++) {
        const d = Math.abs(j.probs[i][k] - r.probs[i][k]);
        maxAbs = Math.max(maxAbs, d);
        if ((j.probs[i][k] > THRESH) !== (r.probs[i][k] > THRESH)) flips.push({ id: r.id, variant: r.variant, pos: i, cls: k ? '。' : '、', js: j.probs[i][k], py: r.probs[i][k] });
      }
    }
    if (j.out !== r.out && diffs.length < 10) diffs.push({ id: r.id, variant: r.variant, py: r.out, js: j.out });
  }
  js[prec] = rows;

  report.precisions[prec] = {
    loadMs: Math.round(loadMs),
    probe: { js: model.probe, py: py.probe },
    rows: py.rows.length,
    textExact_vsPython: rate(exact, py.rows.length),
    rawExact_vsPython: rate(rawExact, py.rows.length),
    skeletonKept: rate(skeleton, py.rows.length),
    probMaxAbsDiff_vsPython: maxAbs,
    thresholdFlips_vsPython: flips,
    diffs,
  };
  await model.release();
}

// int8 vs fp32, on each runtime, at the text and at the decision level.
const agree = (a, b) => a.filter((x, i) => x.out === b[i].out).length / a.length;
const decisionAgree = (a, b) => a.reduce((acc, x, i) => {
  const m = markAgreement(decisions(x.probs), decisions(b[i].probs));
  for (const k of Object.keys(m)) acc[k] = (acc[k] ?? 0) + m[k];
  return acc;
}, {});
report.int8VsFp32 = {
  js: { textExact: agree(js.int8, js.fp32), marks_int8_vs_fp32: decisionAgree(js.int8, js.fp32) },
  python: { textExact: ref.int8_vs_fp32_exact, marks_int8_vs_fp32: decisionAgree(ref.precisions.int8.rows, ref.precisions.fp32.rows) },
  diffs: js.int8.map((x, i) => (x.out !== js.fp32[i].out ? { int8: x.out, fp32: js.fp32[i].out } : null)).filter(Boolean).slice(0, 10),
};

await writeFile(join(root, 'results', 'parity-mojicast.json'), JSON.stringify(report, null, 1));
for (const [prec, p] of Object.entries(report.precisions)) {
  console.log(prec, JSON.stringify({ ...p, thresholdFlips_vsPython: p.thresholdFlips_vsPython.length, diffs: p.diffs.length }, null, 1));
  for (const f of p.thresholdFlips_vsPython.slice(0, 10)) console.log('  FLIP', JSON.stringify(f));
  for (const d of p.diffs) console.log('  DIFF', JSON.stringify(d));
}
console.log('int8 vs fp32', JSON.stringify(report.int8VsFp32, null, 1));
console.log('wrote results/parity-mojicast.json');
